apm-sniffer/optional-plugins/exception-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/exception/ignore/ExceptionIgnorerService.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.exception.ignore;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.skywalking.apm.agent.core.boot.OverrideImplementor;
+import org.apache.skywalking.apm.agent.core.context.DefaultExceptionIgnorerService;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractTracingSpan;
+import org.apache.skywalking.apm.agent.core.plugin.PluginBootstrap;
+import org.apache.skywalking.apm.toolkit.trace.IgnoredException;
+import org.apache.skywalking.apm.util.StringUtil;
+
+import static org.apache.skywalking.apm.plugin.exception.ignore.ExceptionAnalyzerConfig.Plugin.IgnoredException.IGNORED_EXCEPTION_CLASS_NAMES;
+import static org.apache.skywalking.apm.plugin.exception.ignore.ExceptionAnalyzerConfig.Plugin.IgnoredException.IGNORED_EXCEPTION_SCAN_SCOPE;
+
+@OverrideImplementor(DefaultExceptionIgnorerService.class)
+public class ExceptionIgnorerService extends DefaultExceptionIgnorerService {
+
+    private final Set<String> exceptionClassNames = new HashSet<>();
+
+    @Override
+    public void prepare() throws Throwable {
+        if (StringUtil.isNotEmpty(IGNORED_EXCEPTION_CLASS_NAMES)) {
+            exceptionClassNames.addAll(Arrays.asList(IGNORED_EXCEPTION_CLASS_NAMES.split(",")));
+        }
+        if (StringUtil.isNotEmpty(IGNORED_EXCEPTION_SCAN_SCOPE)) {
+            ClassPath classpath = ClassPath.from(PluginBootstrap.class.getClassLoader());
+            ImmutableSet<ClassPath.ClassInfo> classes = classpath.getTopLevelClasses(IGNORED_EXCEPTION_SCAN_SCOPE);
```

Comment:
I've done this before in self apm system,there is some thinking about this feature.

1. the package classes scan is only for the exceptions tagged with @IgnoredException in  user code,so we don't need to scan all classes.PluginBootstrap.class.getClassLoader() have been meet the demand.
2. If some exception defined in other lib which out of user control, the exceptions cannot be tagged be @IgnoredException,so package scan scope is no sense to it. the kind of exception should be ignored by hard code config.

In my mind , user would be like to add a new  ignored exception by @IgnoredException annotation rather than change skywalking agent config frequency.

How do you think about it? @wu-sheng

Reply:
> In my mind , user would be like to add a new ignored exception by @IgnoredException annotation rather than change skywalking agent config frequency.

I don't agree, for those exceptions thrown by some frameworks, users cannot do anything to ignore them but only change the agent config, and I suppose that's a common case where users want to ignore exceptions from frameworks, if they have control to where the exceptions are thrown, they can choose not to throw the exceptions and only log some errors

Reply:
> I don't agree, for those exceptions thrown by some frameworks, users cannot do anything to ignore them but only change the agent config

yes, that is the 2 point what i thought. 
> 2\. If some exception defined in other lib which out of user control, the exceptions cannot be tagged be @IgnoredException,so package scan scope is no sense to it. the kind of exception should be ignored by hard code config.

my option is @IgnoredException is more friendly to use in user service code.





> 1. the package classes scan is only for the exceptions tagged with @IgnoredException in  user code,so we don't need to scan all classes.PluginBootstrap.class.getClassLoader() have been meet the demand.

Reply:
> the package classes scan is only for the exceptions tagged with @IgnoredException in user code,so we don't need to scan all classes.PluginBootstrap.class.getClassLoader() have been meet the demand.

That would be the case. Once you open `@IgnoredException` you can't tell where it will be used. `PluginBootstrap.class.getClassLoader()` wouldn't be able to access many classes when you use the standard J2EE container. Scan tech is not allowed in the agent core stack or plugin stack.

Reply:
`User code` is the wrong statement. All codes are user codes from the agent's perspective. And the user could have a class loader, the framework could have a class loader.

Reply:
this would be cause so many problems.

1.  PluginBootstrap.class.getClassLoader() is application classloader, cannot access custom classloder classes.
2. So we cannot to solve it but for to judge it at run time,for excample have following steps.
     2.1 find classes in every call by the hard config
     2.2 check the status
I don't know if this has a performance impact when find classes instance every time

@wu-sheng @kezhenxu94 
Do you have any ideas to implement this feature?

Reply:
That is why, I asked you don't do the scan for class match. You should determine the flow after exception happens and logs.

Reply:
> 1. PluginBootstrap.class.getClassLoader() is application classloader, cannot access custom classloder classes.

If you check the exception only when it's thrown(in the `Span#log(Throwable e)`), you can simply use the same class loader to load the configured class by name

> 2. So we cannot to solve it but for to judge it at run time,for excample have following steps.
>    2.1 find classes in every call by the hard config
>    2.2 check the status
>    I don't know if this has a performance impact when find classes instance every time

Some simple caching is enough, I don't think the `Ignored Exception` list would be very long

Reply:
it's almost similar to a custom implement skywalking plugin in our company, i would push it laster,thx

> Some simple caching should be enough, I don't think the `Ignored Exception` list would be very long